File: packages/ts/src/authz/relation-enum.ts
/**
 * defineRelations — declarative Zanzibar relation definitions.
 *
 * TypeScript-idiomatic alternative to Python's RelationEnum.
 *
 * Usage:
 *   const Rels = defineRelations({
 *     DOC_OWNER: "doc#owner",
 *     DOC_VIEWER: ["doc", "viewer"],
 *     FOLDER_EDITOR: new Relation("folder", "editor"),
 *   });
 *   Rels.DOC_OWNER            // Relation instance
 *   Rels.DOC_OWNER.toString() // "doc#owner"
 *   Rels.DOC_OWNER.tuple("readme", "user:alice") // RelationTuple
 */

import { Relation, type RelationParser } from "./primitives";

type RelationDef = string | [string, string] | Relation;
type RelationDefs = Record<string, RelationDef>;

type RelationMap<T extends RelationDefs> = {
  readonly [K in keyof T]: Relation;
};

/** Create a frozen map of named Relation instances. */
export function defineRelations<T extends RelationDefs>(
  defs: T,
  options?: { parser?: RelationParser },
): RelationMap<T> {
  const result = {} as Record<string, Relation>;
  for (const [key, def] of Object.entries(defs)) {
    if (def instanceof Relation) {
      result[key] = def;
    } else if (Array.isArray(def)) {
      result[key] = new Relation(def[0], def[1]);
    } else {
      result[key] = new Relation(def, undefined, { parser: options?.parser });
    }
  }
  return Object.freeze(result) as RelationMap<T>;
}

File: packages/py/src/urauth/authz/relation_enum.py
"""RelationEnum — declarative Zanzibar relation definitions using enums."""

from __future__ import annotations

from enum import Enum
from typing import Any

from .primitives import Relation, RelationTuple


class RelationEnum(Enum):
    """Base class for declarative relation definitions.

    Each member's ``.value`` is a ``Relation`` instance (resource-first).
    Members delegate ``__str__``, ``__eq__``, ``__hash__`` to their value.

    Separator is auto-detected from the string form::

        class Rels(RelationEnum):
            DOC_OWNER = "doc#owner"                        # hash separator
            DOC_VIEWER = ("doc", "viewer")                 # tuple form
            FOLDER_EDITOR = Relation("folder", "editor")   # Relation object

    For edge cases, define ``__parser__`` to override auto-detection::

        class Rels(RelationEnum):
            __parser__ = my_custom_parser
            CUSTOM = "some:custom:format"
    """

    def __new__(cls, *args: Any) -> RelationEnum:
        parser = cls.__dict__.get("__parser__")

        if len(args) == 2 and isinstance(args[0], str) and isinstance(args[1], str):
            # Tuple form: ("doc", "owner") → Relation("doc", "owner")
            rel = Relation(str(args[0]), str(args[1]))
        elif len(args) == 1 and isinstance(args[0], str):
            # String form: "doc#owner" → auto-detect separator
            rel = Relation(args[0], parser=parser)
        elif len(args) == 1 and isinstance(args[0], Relation):
            rel = args[0]
        else:
            raise TypeError(
                f"RelationEnum requires a string, (resource, name) tuple, or Relation, got {args!r}"
            )

        obj = object.__new__(cls)
        obj._value_ = rel
        return obj

    def tuple(self, object_id: str, subject: str | None = None) -> RelationTuple:
        """Create a full Zanzibar tuple from this relation definition."""
        return self.value.tuple(object_id, subject)

    def __str__(self) -> str:
        return str(self.value)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, RelationEnum):
            return self.value == other.value
        if isinstance(other, (Relation, str)):
            return self.value == other
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self.value)
